src/OrleansDockerUtils/DockerMembershipOracle.cs

```diff
@@ -0,0 +1,468 @@
+                }
+
+                // Identify removed silos.
+                foreach (var silo in silos)
+                {
+                    // Never remove self.
+                    if (silo.Key.Equals(SiloAddress)) continue;
+
+                    // Do not remove silos which were present in the update.
+                    if (silo.Value.Refreshed)
+                    {
+                        // Reset the refresh flag.
+                        silo.Value.Refreshed = false;
+                        continue;
+                    }
+
+                    // The silo was not present in the update, remove it.
+                    if (removed == null) removed = new HashSet<SiloAddress>();
+                    removed.Add(silo.Key);
+                }
+
+                // If any silos were removed, 
+                if (removed != null)
+                {
+                    foreach (var silo in removed)
+                    {
+                        silos.Remove(silo);
+                    }
+                }
+            }
+
+            // If anything was updated, clear the cache before notifying clients.
+            if (added != null || removed != null)
+            {
+                // Clear the caches.
+                ClearCaches();
+
+                // If any silos were added or removed, notify subscribers of the new status.
+                var siloStatusListeners = subscribers.Values.ToList();
+                if (added != null)
+                {
+                    foreach (var address in added)
+                    {
+                        NotifySubscribers(address, SiloStatus.Active, siloStatusListeners);
+                    }
+                }
+
+                if (removed != null)
+                {
+                    foreach (var address in removed)
+                    {
+                        NotifySubscribers(address, SiloStatus.Dead, siloStatusListeners);
+                    }
+                }
+            }
+        }
+
+        private async Task RefreshAsync()
+        {
+            try
+            {
+                await resolver.Refresh();
+
+                lastRefreshTime = DateTime.UtcNow;
+            }
+            catch (Exception exception)
+            {
+                log?.Warn(
+                    (int)ErrorCode.Docker_MembershipOracle_ExceptionRefreshingSilos,
+                    "Exception refreshing silos.",
+                    exception);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Clears the cached data.
+        /// </summary>
+        private void ClearCaches()
+        {
+            activeSilosCache = null;
+            allSilosCache = null;
+            multiClusterSilosCache = null;
+        }
+
+        /// <summary>
+        /// Updates the status of this silo.
+        /// </summary>
+        /// <param name="status">The updated status.</param>
+        private void UpdateStatus(SiloStatus status)
```

Comment:
This is a somewhat critical question I think so it may deserve its own issue.
If we only need 2 states for SF livenes, then maybe we need only 2 for the regular case as well? Maybe we can simplify the silo state machine and get rid of all other states and stay with just 2?
One need to look at what we are getting out if the other states. I wonder what @sergetbykov thinks. Would like to hear his opinion.

One thing is sure in my eyes: we should try to have similar silo state machines in all livenes cases, and not just ditch states in some cases cause we can't implementat them. If u go that route, it may work, or more likely you will be supporting diverging implementaion and at the end paying higher cost.

Alternatively: when I wrote the first integration with SF(WF at a time) 3-4 years ago, it was similar to your code now, and in addition my plan was to use their build in key value store to store silo states. I didn't implemenent it, but maybe it can work.

 If u go that route, them yet another alternative is to ditch SF livenes and just use the key value store to implement MBR table. That is exactly what we did for Consul. It works, so here is an advantage. And Consul is comparable to SF, so why not? More choices.:-)